AlignPlugin

spetex opened the pull request:
Add align plugin for left, right, and center alignments of elements and marks.

It includes Plugin itself and withAlign() function as well.

spetex commented on slate-plugins.code-workspace:
```diff
@@ -3,10 +3,6 @@
-    },
-    {
-      "name": "slate-plugins",
-      "path": "./packages/slate-plugins"
```
Hmm this is probably done by the editor, I will revert that.

spetex commented on packages/slate-plugins/src/elements/align/components/ToolbarAlign.tsx:
```diff
@@ -0,0 +1,27 @@
+      active={isNodeInSelection(editor, type)}
+      onMouseDown={(event) => {
+        event.preventDefault();
+        console.log({ type });
```
Ooops

spetex commented:
@zbeyens I cannot seem to figure out the problem with dependency cycles, ani insights?

zbeyens commented on packages/slate-plugins/src/elements/align/types.ts:
```diff
@@ -0,0 +1,43 @@
+import { RenderElementProps } from 'slate-react';
+import { RenderElementOptions } from '../../element';
+
+const LEFT = 'left';
+const CENTER = 'center';
+const RIGHT = 'right';
```
`ALIGN_LEFT =  'align_left'` would be more explicit

spetex replied:
I can do `ALIGN_LEFT = 'left'` because the value is used for the CSS attribute.

zbeyens commented on packages/slate-plugins/src/elements/align/transforms/wrapAlign.ts:
```diff
@@ -0,0 +1,15 @@
+import { Editor, Transforms } from 'slate';
+import { unwrapNodesByType } from '../../../common/transforms';
+import { CENTER, LEFT, RIGHT } from '..';
```
```suggestion
import { CENTER, LEFT, RIGHT } from '../types';
```

zbeyens commented on packages/slate-plugins/src/elements/align/renderElementAlign.ts:
```diff
@@ -0,0 +1,10 @@
+import { getRenderElements } from '../../element/utils';
+import { AlignElement } from './components/AlignElement';
```
```suggestion
import { AlignElement } from './components';
```

zbeyens commented on packages/slate-plugins/src/elements/align/unwrapAlign.ts:
```diff
@@ -0,0 +1,19 @@
+
+export const unwrapAlign = (
+  editor = Editor,
+  { LEFT = left, CENTER = center, RIGHT = right } = {}
```
```suggestion
  { typeAlignLeft = LEFT, typeAlignCenter = CENTER, typeAlignRight = RIGHT } = {}
```

zbeyens commented on packages/slate-plugins/src/elements/align/unwrapAlign.ts:
```diff
@@ -0,0 +1,19 @@
+import { castArray } from 'lodash';
+import { Editor, Transforms } from 'slate';
+import { CENTER as center, LEFT as left, RIGHT as right } from './types';
```
```suggestion
import { CENTER, LEFT, RIGHT } from './types';
```

zbeyens commented on packages/slate-plugins/src/elements/align/unwrapAlign.ts:
```diff
@@ -0,0 +1,19 @@
+import { Editor, Transforms } from 'slate';
+import { CENTER as center, LEFT as left, RIGHT as right } from './types';
+
+export const unwrapNodesByType = (editor: any, types: any, options = {}) => {
```
Duplicate?

spetex replied:
yep, I have missed that

zbeyens commented on packages/slate-plugins/src/elements/align/withAlign.ts:
```diff
@@ -0,0 +1,25 @@
+  const { insertData, insertText } = editor;
+
+  editor.insertText = (text) => {
+    if (text) {
+      wrapAlign(editor, text);
+    } else {
```
Why? 🙂

spetex replied:
I got 'inspired' by the LinkPlugin, this might be the result :-P

spetex commented:
I have fixed all of the comments. I have removed the withAlign as it seemed to me it is not necessary. However, I think I have failed to understand its reason.

zbeyens commented on packages/slate-plugins/src/elements/align/deserializeAlign.ts:
```diff
@@ -0,0 +1,35 @@
+import { DeserializeHtml } from '../../common';
+import { AlignDeserializeOptions, CENTER, LEFT, RIGHT } from './types';
+
+const getNodeDeserializer = (
+  type: string,
+  { createNode, tagNames }: { createNode: Function; tagNames: string[] }
+) => {
+  const deserializer = {
+    [type]: createNode,
+  };
+  return tagNames.reduce((obj: any, tagName: string) => {
+    obj[tagName] = createNode;
+    return obj;
+  }, deserializer);
+};
+
+const getElementDeserializer = (
+  type: string,
+  {
+    createElement = () => ({ type }),
+    tagNames = [],
+  }: { createElement?: Function; tagNames: string[] }
+) => getNodeDeserializer(type, { createNode: createElement, tagNames });
```
Duplicate

spetex replied:
Removed

spetex commented:
Rebased on top of next and fixed for compatibility with the `core` refactor. I believe this is ready to merge and release. @zbeyens

zbeyens commented:
Great! Gonna create a release tonight
